### PR TITLE
Avoid using execution_space as View template argument

### LINKED ISSUE #3475: Can't specify memory space for internal Bitset in UnorderedMap construction
When we create a  Kokkos::UnorderedMap, an internal Bitset is created using the execution space of the UnorderedMap. This means that the Bitset allocation uses the default memory space nested in that execution space. This is normally not an issue unless the execution space is Cuda and the default memory space is configured to be UVM (in Trilinos). We are trying to make all of Trilinos libraries UVM free but if applications are counting on UVM and using the Default space on Cuda, they might configure UVM default on. Can the Bitset and UnorderedMap be modified to not use the default space for allocation?

As an example, I hacked the memory events profile tool to show the allocation spaces for a phalanx test where trilinos is configured with UVM enabled. You can see that the Bitset in the UnorderedMap is still using UVM. If I reconfigure Trilinos with UVM disabled then everything is allocated in non-UVM CudaSpace.
```
SpaceHandle=Host, allocation name=grad_basis_ref_mirror
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=CudaUVM, allocation name=Bitset
SpaceHandle=Cuda, allocation name=UnorderedMap hash list
SpaceHandle=Cuda, allocation name=UnorderedMap next index
SpaceHandle=Cuda, allocation name=UnorderedMap keys
SpaceHandle=Cuda, allocation name=UnorderedMap values
SpaceHandle=Cuda, allocation name=UnorderedMap scalars
SpaceHandle=Cuda, allocation name=row_count
SpaceHandle=Cuda, allocation name=row_offsets
SpaceHandle=Cuda, allocation name=matrix_size
SpaceHandle=CudaUVM, allocation name=Bitset
SpaceHandle=Cuda, allocation name=UnorderedMap hash list
SpaceHandle=Cuda, allocation name=UnorderedMap next index
SpaceHandle=Cuda, allocation name=UnorderedMap keys
SpaceHandle=Cuda, allocation name=UnorderedMap values
SpaceHandle=Cuda, allocation name=UnorderedMap scalars
SpaceHandle=Cuda, allocation name=graph_entries
```

Note - the phalanx test code above was refactored to not use the default mem space anymore. This has not been pushed into Trilinos yet, but should be added soon. 

Also - I realize you can still use UVM with the configure flag for UVM disabled. The configure flag only toggles the default memory space. Unfortunately many applications rely on this functionality.

### PR DESCRIPTION
**Pull request by masterleinad**
Should also fix #3475.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Review by crtrott — APPROVED** [review]

**Comment by masterleinad** [comment]
> Is this still a work in progress or is it ready for review?

Seems like it worked immediately. :slightly_smiling_face:

**Comment by dalg24** [comment]
Also this is not adding any test.  Did you verify it resolves the issue reported?

**Comment by masterleinad** [comment]
> Also this is not adding any test. How did you verify it resolves the issue reported?

I checked manually compiling with `UVM` and then setting the memory_space for `UnorderedMap` to `CudaSpace` in the `UnorderedMap` unit tests that we don't allocate in `UVM` space. I didn't check anything with respect to the other changes.

I'm not quite sure if we require a test here.

**Review by crtrott — CHANGES_REQUESTED** [review]
After thinking it more over I agree that DeviceType is better as the type than DeviceType::memory_space, it actually is closer to the original intend of using the optimal execution memory space for the chosen execution space.

File: algorithms/src/Kokkos_Random.hpp [hunk 1]
```diff
@@ -1140,7 +1140,7 @@ template <class DeviceType = Kokkos::DefaultExecutionSpace>
 class Random_XorShift1024_Pool {
  private:
   using execution_space = typename DeviceType::execution_space;
-  using locks_type      = View<int**, execution_space>;
+  using locks_type      = View<int**, typename DeviceType::memory_space>;
```
Comment (dalg24) [inline algorithms/src/Kokkos_Random.hpp:1143]:
```suggestion
  using locks_type      = View<int**, DeviceType>;
```

File: core/src/impl/Kokkos_ViewArray.hpp [hunk 2]
```diff
@@ -355,7 +355,9 @@ class ViewMapping<Traits, Kokkos::Array<>> {
 
     using execution_space = typename alloc_prop::execution_space;
     using memory_space    = typename Traits::memory_space;
-    using functor_type    = ViewValueFunctor<execution_space, scalar_type>;
+    using functor_type =
+        ViewValueFunctor<Kokkos::Device<execution_space, memory_space>,
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_ViewArray.hpp:359]: PIng

[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/impl/Kokkos_ViewArray.hpp]: See 820c00cf9.

**Review by crtrott — APPROVED** [review]
After thinking it more over I agree that DeviceType is better as the type than DeviceType::memory_space, it actually is closer to the original intend of using the optimal execution memory space for the chosen execution space.

**Pull request merged**